### PR TITLE
[JumpThreading] Thread over BB with only an unconditional branch

### LINKED ISSUE #76609: [JumpThreading] Missing optimization: thread over a nearly empty basicblock
Alive2 proof: https://alive2.llvm.org/ce/z/Lv9e4r

### Description:

```llvm
define i1 @src(i1 %c1) {
entry:                                      
  %c = call i1 @cond()
  br i1 %c, label %then, label %else

then:                                        
  call void @dummy()
  br i1 %c1, label %else, label %return

else:                                         
  br label %return

return:                                           
  %retval.0 = phi i1 [ true, %else ], [ false, %then ]
  ret i1 %retval.0
}
```
can be folded to:
```llvm
define i1 @tgt(i1 %c1) {
entry:
  %c = call i1 @cond()
  br i1 %c, label %then, label %return

then:                                        
  call void @dummy()
  br i1 %c1, label %else, label %return

else:                                         
  br label %return

return:                                           
  %retval.0 = phi i1 [ true, %entry ], [ false, %then ], [true, %else]
  ret i1 %retval.0
}
```
The conditional branch in BB `entry`  can thread directly through nearly empty BB `else` to `return`.

### Real-world motivation

This snippet of IR is derived from auto-generated [qemu/xxx/qapi-visit-dump.c](https://github.com/qemu/qemu/blob/master/qapi/dump.json) (after inline and O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, contact me to get it(Too big to upload here).

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by XChy**
Fixes #76609
This patch does:
- relax the phis constraint in `CanRedirectPredsOfEmptyBBToSucc`
- guarantee the BB has multiple different predecessors to redirect, so that we can handle the case without phis in BB. Without this change and phi constraint, we may redirect the CommonPred.

The motivation is consistent with JumpThreading. We always want the branch to jump more direct to the destination, without passing the middle block. In this way, we can expose more other optimization opportunities.

An obivous example proposed by @dtcxzyw is like:
```llvm
define i32 @test(...) {
entry:
   br i1 %c, label %do.end, label %if.then

if.then:                                          ; preds = %entry
   %call2 = call i32 @dummy()
   %tobool3.not = icmp eq i32 %call2, 0
   br i1 %tobool3.not, label %do.end, label %return

do.end:                                           ; preds = %entry, %if.then
   br label %return

return:                                           ; preds = %if.then, %do.end
   %retval.0 = phi i32 [ 0, %do.end ], [ %call2, %if.then ]
   ret i32 %retval.0
}
```
`entry` can directly jump to return, without passing `do.end`, and then the if-else pattern can be simplified further:
```llvm
define i32 @test(...) {
entry:
   br i1 %c, label %return, label %if.then

if.then:                                          ; preds = %entry
   %call2 = call i32 @dummy()
   br label %return

return:                                           ; preds = %if.then
   %retval.0 = phi i32 [ 0, %entry ], [ %call2, %if.then ]
   ret i32 %retval.0
}
```

### CODE REVIEW

File: llvm/test/Transforms/JumpThreading/uncond-no-phi.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/JumpThreading/uncond-no-phi.ll:11]: Could you please add some tests to demonstrate that this patch will save some instructions?

An example:
```
define i32 @test(...) {
   %tobool.not = icmp eq ptr %0, null
   br i1 %tobool.not, label %do.end, label %if.then

 if.then:                                          ; preds = %entry
   %call2 = tail call i32 %visit(ptr noundef nonnull %0, ptr noundef %arg) #6
   %tobool3.not = icmp eq i32 %call2, 0
   br i1 %tobool3.not, label %do.end, label %return

 do.end:                                           ; preds = %entry, %if.then
   br label %return

 return:                                           ; preds = %if.then, %do.end
   %retval.0 = phi i32 [ 0, %do.end ], [ %call2, %if.then ]
   ret i32 %retval.0
}
```
It will be folded into:
```
define i32 @test(...) {
   %tobool.not = icmp eq ptr %0, null
   br i1 %tobool.not, label %return, label %if.then

 if.then:                                          ; preds = %entry
   %call2 = tail call i32 %visit(ptr noundef nonnull %0, ptr noundef %arg) #6
   br label %return

 return:                                           ; preds = %if.then, %do.end
   %retval.0 = phi i32 [ 0, %entry ], [ %call2, %if.then ]
   ret i32 %retval.0
}
```

File: llvm/lib/Transforms/Utils/Local.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/Local.cpp:1017]: I think the original comment is easier to understand than this one.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

File: llvm/test/Transforms/JumpThreading/uncond-no-phi.ll
Comment (DianQK) [reply llvm/test/Transforms/JumpThreading/uncond-no-phi.ll:11]: I would suggest adding it to the PR description as well. This makes it easier for people to know why we're doing this conversion that doesn't seem to be an improvement. I think it's also worth adding a test case in `PhaseOrder`.

Comment (XChy) [reply llvm/test/Transforms/JumpThreading/uncond-no-phi.ll:11]: I will add the tests to SimplifyCFG, which also call the same function to thread it.

File: llvm/test/Transforms/SimplifyCFG/branch-fold.ll
Comment (DianQK) [inline llvm/test/Transforms/SimplifyCFG/branch-fold.ll:9]: Please discard this change.

File: llvm/test/CodeGen/AArch64/and-sink.ll
Comment (DianQK) [inline llvm/test/CodeGen/AArch64/and-sink.ll:1]: Please submit this change separately.
I'm not sure if this test case is broken.

File: llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll
Comment (DianQK) [inline llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll:34]: I think it should be put in `llvm/test/Transforms/PhaseOrdering`. I'm curious that why you added two similar test cases. Please add comments if necessary.

Comment (XChy) [reply llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll:34]: I think that the phase-ordering tests emphasize the impact after changing pipeline. I don't have a strong opinion on this, update it if you're more happy with PhaseOrdering.
I add two similar testcases because they are found in the benchmark frequently. And the icmp of one of them is not optimized away, it could be seen as not-that-optimal one.

Comment (DianQK) [reply llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll:34]: > I think that the phase-ordering tests emphasize the impact after changing pipeline. I don't have a strong opinion on this, update it if you're more happy with PhaseOrdering.

IIUC, we often put test cases that require multiple passes here. :)

> I add two similar testcases because they are found in the benchmark frequently. And the icmp of one of them is not optimized away, it could be seen as not-that-optimal one.

IIUC, this cannot be optimized.

Comment (XChy) [reply llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll]: The tests have been moved to `PhaseOrdering`.
> IIUC, this cannot be optimized.

Yes, it cannot be optimized away. I just meant this patch optimizes it, but the effect is not that obvious as optimizing the first test. Do you have some suggestions on it?

Comment (DianQK) [reply llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll]: > The tests have been moved to `PhaseOrdering`.
> 
> > IIUC, this cannot be optimized.
> 
> Yes, it cannot be optimized away. I just meant this patch optimizes it, but the effect is not that obvious as optimizing the first test. Do you have some suggestions on it?

It seems like an improvement: https://llvm.godbolt.org/z/3hqjbehvW.

**Comment by DianQK** [comment]
Hmm, please consider the following case: 

```llvm
define i32 @thread_uncond_bb_cmp_zext(i1 %c, i32 %v, i32 %v2, i32 %v3) {
entry:
  br i1 %c, label %do.end, label %if.then

if.then:                                          ; preds = %entry
  call void @dummy()
  %tobool = icmp eq i32 %v, 0
  br i1 %tobool, label %do.end, label %return

do.end:                                           ; preds = %entry, %if.then
  br label %return

return:                                           ; preds = %if.then, %do.end
  %retval = phi i32 [ %v2, %do.end ], [ %v3, %if.then ]
  ret i32 %retval
}

declare void @dummy()
```

https://llvm.godbolt.org/z/74f7qz54E.

The generated code will become more complex if other passes cannot eliminate a branch instruction or convert it into a select instruction.

File: llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll
Comment (DianQK) [reply llvm/test/Transforms/SimplifyCFG/thread-uncond-bb.ll]: I think we should keep this case. The test cases in `PhaseOrdering` should ensure that such regressions do not occur again.

**Comment by XChy** [comment]
@DianQK, in most cases, we can produce a select instruction for the branch. And whether fold it to select in generated code depends on the backend. The armv7 and aarch64 backends emit the select as expected for your example: https://llvm.godbolt.org/z/v4ParEM8E

**Review by dianqk — COMMENTED** [review]
LGTM with "Comment" :3. (It's usually just because I haven't prepared myself to approve.)

My main concern is the impact of introducing more phi nodes in larger programs. Perhaps I'm overthinking it.
~~If I remember correctly, #85510 is also affected because `CanRedirectPredsOfEmptyBBToSucc` can only see less facts.~~

**Review by dianqk — APPROVED** [review]

**Comment by XChy** [comment]
I would like to merge this, @nikic , any comment?

File: llvm/test/Transforms/JumpThreading/select.ll
Comment (nikic) [inline llvm/test/Transforms/JumpThreading/select.ll:24]: Why is this CHECK-BPI now?

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/JumpThreading/select.ll
Comment (XChy) [reply llvm/test/Transforms/JumpThreading/select.ll:24]: I have no idea about it, `update_test_checks` replaces it even without this patch. Seems the default behavior of multiple prefixes changes for globals.

Comment (nikic) [reply llvm/test/Transforms/JumpThreading/select.ll:24]: It looks like this is somehow related to the debug output. I think you can modify the RUN lines to look like this:
```
; RUN: opt -S -passes="jump-threading" < %s 2>&1 | FileCheck %s
; RUN: opt -S -passes="require<branch-prob>,jump-threading" -debug-only=branch-prob -disable-output < %s 2>&1 | FileCheck -check-prefix=CHECK-BPI %s
```

Comment (XChy) [reply llvm/test/Transforms/JumpThreading/select.ll]: Thanks, it works now.

**Comment by nikic** [comment]
This seems to cause a 0.15% compile-time regression using stage2 clang only (https://llvm-compile-time-tracker.com/compare.php?from=1120d8e6f799121b611aa23bdc128e40cf9c6c58&to=36b3c26451bf9a42f0b6b415993d3942bb73abdd&stat=instructions:u), which indicates that causes an optimization regression, at least for clang :(

**Comment by XChy** [comment]
Emm, unexpected result :(. Since I can't see obvious regression in *llvm-opt-benchmark* and have no simple regression testcase for it, I can't fix it immediately. If emergent, feel free to revert this patch.

**Comment by dyung** [comment]
@XChy your change seems to be causing a test failure of `llvm/test/CodeGen/Generic/live-debug-label.ll` on several buildbots:
- https://lab.llvm.org/buildbot/#/builders/183/builds/20800
- https://lab.llvm.org/buildbot/#/builders/188/builds/44495
- https://lab.llvm.org/buildbot/#/builders/245/builds/23249
- https://lab.llvm.org/buildbot/#/builders/280/builds/1949
- https://lab.llvm.org/buildbot/#/builders/119/builds/17315

Can you take a look and revert if you need time to investigate?

**Comment by XChy** [comment]
For compile-time regression and a failure on debug info test(I can't reproduce it on my machine), revert this patch now.

**Pull request merged**
